File: backend/app/schemas/task.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, field_validator

from app.models.task import TaskPriority


class TaskCreate(BaseModel):
    title: str
    description: str | None = None
    priority: TaskPriority = TaskPriority.WANT
    total_duration_minutes: int
    min_block_minutes: int = 30
    max_block_minutes: int = 120
    deadline: date | None = None
    window_start: date | None = None
    window_end: date | None = None
    recurrence_period_days: int | None = None

    @field_validator("total_duration_minutes", "min_block_minutes", "max_block_minutes")
    @classmethod
    def must_be_positive(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("Duration must be positive")
        return v


class TaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    priority: TaskPriority | None = None
    total_duration_minutes: int | None = None
    min_block_minutes: int | None = None
    max_block_minutes: int | None = None
    deadline: date | None = None
    window_start: date | None = None
    window_end: date | None = None
    recurrence_period_days: int | None = None
    completion_percentage: float | None = None
    is_complete: bool | None = None


class TaskResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    description: str | None
    priority: str
    total_duration_minutes: int
    min_block_minutes: int
    max_block_minutes: int
    deadline: date | None
    window_start: date | None
    window_end: date | None
    recurrence_period_days: int | None
    completion_percentage: float
    is_complete: bool
    source: str
    source_id: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: mobile/lib/models/task.dart
class Task {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final String priority; // need | want | like
  final int totalDurationMinutes;
  final int minBlockMinutes;
  final int maxBlockMinutes;
  final String? deadline;
  final double completionPercentage;
  final bool isComplete;
  final String source;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Task({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    required this.priority,
    required this.totalDurationMinutes,
    required this.minBlockMinutes,
    required this.maxBlockMinutes,
    this.deadline,
    required this.completionPercentage,
    required this.isComplete,
    required this.source,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Task.fromJson(Map<String, dynamic> json) => Task(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        title: json['title'] as String,
        description: json['description'] as String?,
        priority: json['priority'] as String,
        totalDurationMinutes: json['total_duration_minutes'] as int,
        minBlockMinutes: json['min_block_minutes'] as int,
        maxBlockMinutes: json['max_block_minutes'] as int,
        deadline: json['deadline'] as String?,
        completionPercentage: (json['completion_percentage'] as num).toDouble(),
        isComplete: json['is_complete'] as bool,
        source: json['source'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: DateTime.parse(json['updated_at'] as String),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'user_id': userId,
        'title': title,
        'description': description,
        'priority': priority,
        'total_duration_minutes': totalDurationMinutes,
        'min_block_minutes': minBlockMinutes,
        'max_block_minutes': maxBlockMinutes,
        'deadline': deadline,
        'completion_percentage': completionPercentage,
        'is_complete': isComplete,
        'source': source,
      };
}
